sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1994,6 +1994,13 @@ object EliminateLimits extends Rule[LogicalPlan] {
       LocalLimit(Literal(Least(Seq(ne, le)).eval().asInstanceOf[Int]), grandChild)
     case Limit(le, Limit(ne, grandChild)) =>
       Limit(Literal(Least(Seq(ne, le)).eval().asInstanceOf[Int]), grandChild)
+
+    case Project(projectList, Limit(l, child)) =>
```

Comment:
`projectList` may contain very heavy UDFs. This change may increase `Project` evaluation cost.

Reply:
Thanks @wangyum for your time. What about:
```scala
case Limit(l, Project(projectList, child)) =>
  Project(projectList, Limit(l, child))
```

Reply:
+1.